### PR TITLE
Add support for URL protocol and file type association (all OS)

### PR DESCRIPTION
**Pull request by aganders3**
### Description
In development of #117 we realized macOS does not support passing URLs via argv when opening from a URL. See also #118 for more discussion.

This draft PR investigates an experimental launcher for the mac bundle that uses AppKit in order to respond to Apple Events, specifically to implement [`application(_:open:)`](https://developer.apple.com/documentation/appkit/nsapplicationdelegate/2887193-application). This allows us to respond to Apple Events associated with opening a custom URL scheme. The launcher first launches the main application via `NSWorkspace` and hides itself, listens for relevant Apple Events and dispatches them to another CLI program, then quits if the main application exits.

The new app bundle structure is a bit more complicated, and is basically the current structure nested as a resource in a new bundle using the proposed launcher.
```
menuinst app name.app
└── Contents
    ├── Entitlements.plist
    ├── Info.plist  <------ this is where we register the url scheme, etc.
    ├── MacOS
    │   └── menuinst-app-name    <----- this is the new Swift launcher from this PR
    ├── PkgInfo
    ├── Resources
    │   ├── menuinst app name.app  <----- this is basically the current app bundle, but now does *not* register the url scheme
    │   ├── menuinst icon.icns
    │   └── open-url    <------ the launcher calls this with the url as arguments whenever a URL is opened
      ...
```

The `open-url` script/program should be somehow user-configurable. Right now it is not included in this PR. Here's a very barebones proof-of-concept that would communicate with a main running app via socket:
```bash
#!/bin/bash

set -ueo pipefail

for i in 1 2 3 4 5
do
    echo "${1#menuinst://}" | nc localhost 40256 && break || sleep 5
done
```

~~There's a known issue where a second dock icon will flash (very) briefly on startup and when opening a URL in the running app.~~

### Checklist
- [ ] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
- [ ] Add / update necessary tests?
- [ ] Add / update outdated documentation?

N.B. @jaimergp

### CODE REVIEW

**Comment by aganders3** [comment]
> There's a known issue where a second dock icon will flash (very) briefly on startup and when opening a URL in the running app.

I believe this is fixed with the latest commit!

**Comment by jaimergp** [comment]
Ok, I added some commits that (a) integrate the now merged #117, and (b) add some tests and examples. The tests currently fail because the association is not working somehow. To be investigated but now we have CI eyes :D

**Comment by jaimergp** [comment]
I think we also need to handle this file ([from SO](https://superuser.com/a/548122)):

> The file you seek is ~/Library/Preferences/com.apple.LaunchServices/com.apple.launchservices.secure.plist.
>
> It holds an array called LSHandlers, and the Dictionary children that define an LSHandlerURLScheme can be modified accordingly with the LSHandlerRole.

File: tests/data/jsons/url_protocols.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+                    "command": [
+                        "{{ PYTHON }}",
+                        "-c",
+                        "import sys, pathlib as p; Path(r'__OUTPUT_FILE__').write_text(sys.argv[1])"
+                    ],
+                    "CFBundleURLTypes": [
+                        {
+                            "CFBundleURLIconFile": "{{ MENU_DIR }}/my_protocol_handler",
+                            "CFBundleURLName": "my-protocol-handler.menuinst",
+                            "CFBundleTypeRole": "Viewer",
```
Comment (aganders3) [inline tests/data/jsons/url_protocols.json:40]: For this scheme to work properly I think the `CFBundleTypeRole` actually needs to be omitted. This took me hours to figure out. I was so desperate to open a thread on Apple's developer forum: https://developer.apple.com/forums/thread/725410

**Comment by jaimergp** [comment]
The file is binary but can be opened with Python's `plistlib`. This is what mine looks like:

```json
{
  "LSHandlers": [
    {
      "LSHandlerURLScheme": "https",
      "LSHandlerRoleAll": "com.google.chrome",
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      }
    },
    {
      "LSHandlerURLScheme": "com.todoist",
      "LSHandlerRoleAll": "com.todoist.mac.todoist",
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      }
    },
    {
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      },
      "LSHandlerRoleAll": "com.microsoft.vscode",
      "LSHandlerContentTagClass": "public.filename-extension",
      "LSHandlerContentTag": "4"
    },
    {
      "LSHandlerURLScheme": "todoist",
      "LSHandlerRoleAll": "com.todoist.mac.todoist",
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      }
    },
    {
      "LSHandlerContentType": "public.html",
      "LSHandlerRoleAll": "com.google.chrome",
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      }
    },
    {
      "LSHandlerURLScheme": "prli",
      "LSHandlerRoleAll": "com.parallels.desktop.console",
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      }
    },
    {
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      },
      "LSHandlerRoleAll": "com.parallels.sharedapplink.agent",
      "LSHandlerContentTagClass": "public.filename-extension",
      "LSHandlerContentTag": "lnk"
    },
    {
      "LSHandlerURLScheme": "webcal",
      "LSHandlerRoleAll": "com.google.chrome",
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      }
    },
    {
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      },
      "LSHandlerRoleAll": "com.microsoft.vscode",
      "LSHandlerContentTagClass": "public.filename-extension",
      "LSHandlerContentTag": "8"
    },
    {
      "LSHandlerURLScheme": "web+stellar",
      "LSHandlerRoleAll": "keybase.electron",
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      }
    },
    {
      "LSHandlerURLScheme": "mailto",
      "LSHandlerRoleAll": "com.google.chrome",
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      }
    },
    {
      "LSHandlerURLScheme": "keybase",
      "LSHandlerRoleAll": "keybase.electron",
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      }
    },
    {
      "LSHandlerURLScheme": "zoomphonecall",
      "LSHandlerRoleAll": "us.zoom.xos",
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      }
    },
    {
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      },
      "LSHandlerRoleAll": "com.microsoft.vscode",
      "LSHandlerContentTagClass": "public.filename-extension",
      "LSHandlerContentTag": "in"
    },
    {
      "LSHandlerURLScheme": "facetime",
      "LSHandlerRoleAll": "com.apple.facetime",
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      }
    },
    {
      "LSHandlerContentType": "public.url",
      "LSHandlerRoleViewer": "com.google.chrome",
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleViewer": "-"
      }
    },
    {
      "LSHandlerURLScheme": "http",
      "LSHandlerRoleAll": "com.google.chrome",
      "LSHandlerPreferredVersions": {
        "LSHandlerRoleAll": "-"
      }
    }
  ]
}
```

**Comment by aganders3** [comment]
Good to know - from my experiments the application is registered automatically when first launched, but it could be useful to directly manipulate this list (e.g. when installing/uninstalling multiple versions).

I think another interface to this file is the `lsregister` command:
```
❮ /System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Support/lsregister --help                          
lsregister: [OPTIONS] [ <path>... ]
                      [ -apps <domain>[,domain]... ]
                      [ -libs <domain>[,domain]... ]
                      [ -all  <domain>[,domain]... ]

Paths are searched for applications to register with the Launch Service database.
Valid domains are "system", "local", "network" and "user". Domains can also
be specified using only the first letter.

  -delete       Delete the Launch Services database file. You must then reboot!
  -kill         Reset the Launch Services database before doing anything else
  -seed         If database isn't seeded, scan default locations for applications and libraries to register
  -lint         Print information about plist errors while registering bundles
  -lazy n       Sleep for n seconds before registering/scanning
  -r            Recursive directory scan, do not recurse into packages or invisible directories
  -R            Recursive directory scan, descending into packages and invisible directories
  -f            force-update registration even if mod date is unchanged
  -u            unregister instead of register
  -v            Display progress information
  -gc           Garbage collect old data and compact the database
  -dump [table] Display full database contents after registration
  -h            Display this help
  ```

**Comment by jaimergp** [comment]
Running that command on the `.app` shortcut doesn't modify the plist database, so I guess something's off with the example I am using. Looking into it.

**Comment by aganders3** [comment]
Or I'm just wrong!

File: tests/data/jsons/url_protocols.json (hunk 1 above)
Comment (jaimergp) [reply tests/data/jsons/url_protocols.json]: Oh, nice, thanks. I added this to the docs.

**Comment by aganders3** [comment]
I'm working on getting the tests to run locally. Shortcut creation (including URL and file registration) is working when I run menuinst manually, but not with the tests. Maybe
LaunchServices is not happy with the .app in a private/temp dir?

**Comment by aganders3** [comment]
Okay it's extra weird because in CI it should skip mocking the path. If I set `CI=true` on my laptop I can get the test to pass.

**Comment by aganders3** [comment]
So the output file is being created but is empty...maybe there is some stream buffering happening due to not being an interactive terminal or something?

**Comment by jaimergp** [comment]
That was very promising and really wanted the newline to do the work for us, but it didn't :( 

At least we know that it works on more machines than mine :D

**Comment by aganders3** [comment]
Thanks! I observed similar but its nice to see this documented so methodically. I will do some research on Apple Events on remote/headless runners.

**Comment by jaimergp** [comment]
Erm... WHY did it pass now? 😂

**Comment by aganders3** [comment]
I was going to suggest perhaps it's prompting for some permissions and hanging? There are a few possibly related issues on the `runner-images` repo, for example https://github.com/actions/runner-images/issues/553.

Now that it passed though...

**Comment by jaimergp** [comment]
Hey @aganders3 @goanpeca, if you have time to take a look here and leave any comments, that'd be appreciated. It ended up being quite the PR, but I am happy with the results. I think it's good enough for a merge to the dev branch!

Thank you!

File: tests/test_api.py [hunk 2]
```diff
@@ -1,80 +1,137 @@
...
+
+    out = ""
+    while not (out := Path(path).read_text()):
+        sleep(1)
+        if time() >= t0 + timeout:
+            raise RuntimeError(f"Timeout. File '{path}' was empty!")
+    return out
+
+
+def check_output_from_shortcut(
```
Comment (aganders3) [inline tests/test_api.py:37]: I think it's okay for now, but with more tests this function becomes pretty tangled. I'd probably move much of the logic into the individual tests even if it means repeating a bit of code.

File: docs/source/defining-shortcuts.md [hunk 3]
```diff
@@ -77,26 +86,39 @@ A multi-platform example:
...
-                    "command": ["{{ PREFIX }}/bin/my_csv_reader.py", "%f"],
-                    "MimeType": ["text/csv"]
+                    "command": ["{{ PREFIX }}/bin/open_menuinst_files.py", "%f"],
+                    "MimeType": ["application/x-menuinst"],
+                    "glob_patterns": {
+                        "application/x-menuinst": ["*.menuinst"]
+                    }
                 },
-                "macos": {
+                "osx": {
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply docs/source/defining-shortcuts.md:99]: The conda ecosystem uses `osx` as a platform prefix (e.g. osx-64, osx-arm64), so it's just for consistency.

File: src/appkit-launcher/Sources/appkit-launcher/appkit-launcher.swift [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    ) { [weak self] (notification: Notification) in
+      if self?.mainApp?.isTerminated ?? false {
+        os_log("Main app finished. Wrapper will terminate", log: self!.osLog, type: .debug)
+        NSApplication.shared.terminate(self)
+      }
+    }
+
+    let env = ProcessInfo.processInfo.environment
+    // TODO: only available on macOS 10.15+, do we need to support older?
+    // https://developer.apple.com/documentation/appkit/nsworkspace/1532940-open
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply src/appkit-launcher/Sources/appkit-launcher/appkit-launcher.swift:74]: Are there any workarounds we can explore?

File: menuinst/platforms/win.py [hunk 5]
```diff
@@ -258,20 +258,35 @@ def _write_script(self, script_path: Optional[os.PathLike] = None) -> Path:
...
-                command = ["cmd", "/K", str(script)]
+                command = ["cmd", "/K", f'"{script}"']
+                if with_arg1:
+                    command.append("%1")
             else:
                 system32 = Path(os.environ.get("SystemRoot", "C:\\Windows")) / "system32"
+                arg1 = "%1 " if with_arg1 else ""
+                # This is an UGLY hack to start the script in a hidden window
+                # We use CMD to call PowerShell to call the BAT file
+                # This flashes faster than Powershell -> BAT! Don't ask me why.
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply menuinst/platforms/win.py:273]: That's a good idea that nonetheless made me shiver 😂

File: menuinst/platforms/osx.py [hunk 6]
```diff
@@ -227,5 +302,21 @@ def _sign_with_entitlements(self):
...
+        needed_keys = ("CFBundleURLTypes", "CFBundleDocumentTypes")
+        return any([self.metadata.get(k) for k in needed_keys])
+
+
+def _lsregister(*args, check=True, **kwargs):
+    exe = (
+        "/System/Library/Frameworks/CoreServices.framework"
+        "/Frameworks/LaunchServices.framework/Support/lsregister"
+    )
+    if not os.path.exists(exe):
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply menuinst/platforms/osx.py:320]: Fair enough.

File: src/appkit-launcher/Sources/appkit-launcher/appkit-launcher.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (aganders3) [reply src/appkit-launcher/Sources/appkit-launcher/appkit-launcher.swift:74]: That link leads to the deprecated APIs the one we're using has replaced:
https://developer.apple.com/documentation/appkit/nsworkspace/1532940-open

So I think we might be able to do something like this? I will need to play around with this to see if it does the same thing.
```swift
if #available(macOS 10.15, *) {
    let config = NSWorkspace.OpenConfiguration()
    config.createsNewApplicationInstance = env.keys.contains("MENUINST_ALLOW_MULTIPLE_INSTANCES")
    config.environment = env

    os_log("Opening app %@", log: self.osLog, type: .debug, self.wrappedApp()!.absoluteString)
    NSWorkspace.shared.openApplication(
      at: self.wrappedApp()!,
      configuration: config,
      completionHandler: { [weak self] app, error in
        self?.mainApp = app
      }
    )
} else {
    self?.mainApp = try! NSWorkspace.launchApplication(
        at: self.wrappedApp()!,
        configuration: [NSWorkspace.LaunchConfigurationKey.environment: environment]
    )
}
```

**Comment by jaimergp** [comment]
I used your snippet and adjusted it a bit so it can compile. One problem with the deprecated version is that the opened process will be left lingering around, which might not be ideal... but at least it will launch something. Hopefully users on <10.15 are not too many :)

File: src/appkit-launcher/Package.swift [hunk 7]
```diff
@@ -0,0 +1,18 @@
+// swift-tools-version: 5.7
+// The swift-tools-version declares the minimum version of Swift required to build this package.
+
+import PackageDescription
+
+let package = Package(
+  name: "appkit-launcher",
+  platforms: [
+    .macOS(.v10_15)
```
Comment (jaimergp) [inline src/appkit-launcher/Package.swift:9]: Check this again for 10.6

**Review by jaimergp — APPROVED** [review]
Let's do this.

**Pull request merged**